### PR TITLE
Add SkoptSampler.

### PR DESCRIPTION
**Pull request by sile**
This implementation is based on the new sampler interface being proposed by https://github.com/pfnet/optuna/pull/380.

### CODE REVIEW

File: optuna/study.py [hunk 1]
```diff
@@ -578,6 +578,41 @@ def trials(self):
 
         return self.storage.get_all_trials(self.study_id)
 
+    @property
+    def product_search_space(self):
+        # type: () -> Dict[str, BaseDistribution]
+        """Return the product search space of the :class:`~optuna.study.InTrialStudy`.
+
+        "product search space" contains the product of parameter distributions that have been
```
Comment (toshihikoyanase) [inline optuna/study.py]: IMO, the expression 'product of parameter distributions' seems to be confusing with [product distribution](https://en.wikipedia.org/wiki/Product_distribution). Can we rephrase this as something like 'product set of parameter distributions'?

File: optuna/integration/skopt.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            If :obj:`None` is specified, :class:`~optuna.samplers.TPESampler` is used
+            as the default. See also :class:`~optuna.samplers`.
+        skopt_kwargs:
+            Keyword arguments passed to the constructor of
+            `skopt.Optimizer <https://scikit-optimize.github.io/#skopt.Optimizer>`_
+            class.
+
+            Note that the ``dimensions`` argument is automatically added by
+            :class:`~optuna.integration.SkoptSampler`, so you don't have to specify it
+            (it will be ignored even if specified).
```
Comment (toshihikoyanase) [inline optuna/integration/skopt.py]: I feel that this message can be simpler. For instance, what do you think of the following sentence?

> Note that ``dimensions`` argument in ``skopt_kwargs`` will be ignored because it is added by :class:`~optuna.integration.SkoptSampler` automatically.

File: optuna/integration/skopt.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    _import_error = e
+    # SkoptSampler is disabled because Scikit-Optimize is not available.
+    _available = False
+
+if types.TYPE_CHECKING:
+    from typing import Any  # NOQA
+    from typing import Dict  # NOQA
+    from typing import List  # NOQA
+    from typing import Optional  # NOQA
+    from typing import Set  # NOQA
```
Comment (toshihikoyanase) [inline optuna/integration/skopt.py]: `Set` is not used in this file. Please remove it.

File: optuna/integration/skopt.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    from typing import Tuple  # NOQA
+    from typing import Union  # NOQA
```
Comment (toshihikoyanase) [inline optuna/integration/skopt.py]: ditto.

File: optuna/integration/skopt.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+                                                            param_distribution)
+
+
+class _Optimizer(object):
+    def __init__(self, search_space, skopt_kwargs=None):
+        # type: (Dict[str, BaseDistribution], Optional[Dict[str, Any]]) -> None
+
+        self._search_space = search_space
+
+        dimensions = []  # type: List[Any]
```
Comment (toshihikoyanase) [inline optuna/integration/skopt.py]: I think the type of `dimensions` can be restricted to `List[space.Dimension]`. Do you have any reason to use `List[Any]`?

File: tests/integration_tests/test_skopt.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            # Original: trial.suggest_discrete_uniform('p8', 2.2, 2.2, 0.5)
+            # => Skipped because `skopt.Optimizer` cannot handle an empty `Real` dimension.
+
+            # Original: trial.suggest_categorical('p9', ['9', '3', '0', '8'])
+            space.Categorical(('9', '3', '0', '8'))
+        ]
+        assert mock_object.mock_calls[0] == call(dimensions)
+
+
+def test_suggested_value():
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_skopt.py]: I think tests for objective functions that returns `NaN` or `None` are missing. Could you add such test cases to confirm `SkoptSampler` handles them correctly?

File: optuna/integration/skopt.py (hunk 3 above)
Comment (sile) [reply optuna/integration/skopt.py]: Thank you for your careful review!

Fixed: https://github.com/pfnet/optuna/pull/399/commits/6598c4c99af4244d829a4bbab1e7fe0695306c72

File: optuna/integration/skopt.py (hunk 4 above)
Comment (sile) [reply optuna/integration/skopt.py]: Fixed: https://github.com/pfnet/optuna/pull/399/commits/6598c4c99af4244d829a4bbab1e7fe0695306c72

File: optuna/integration/skopt.py (hunk 5 above)
Comment (sile) [reply optuna/integration/skopt.py]: `List[Any]` was necessary in the previous code, but it has become useless, so I removed the type hint completely at the commit https://github.com/pfnet/optuna/pull/399/commits/36b7c7f2a86a5f2095b22c3868a31c7a2f1a2e28.

File: optuna/integration/skopt.py (hunk 2 above)
Comment (sile) [reply optuna/integration/skopt.py]: Seems good, thank you!

Fixed: https://github.com/pfnet/optuna/pull/399/commits/e0ca9ac44bb8dedf38bdcb2d6d73afcb5c99da7e

File: tests/integration_tests/test_skopt.py (hunk 6 above)
Comment (sile) [reply tests/integration_tests/test_skopt.py]: I will add tests for `NaN` and `None`. Thank you for pointing it out.

Comment (sile) [reply tests/integration_tests/test_skopt.py]: Added: https://github.com/pfnet/optuna/pull/399/commits/86f9f83c90ea195426e9d0e65e0f21df267aed4e

Because the type of objective function doesn't allow that the function returns `None`, I omitted a test case for `None` return value.

File: optuna/integration/skopt.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        param_values = self._optimizer.ask()
+        for (name, distribution), value in zip(sorted(self._search_space.items()), param_values):
+            if isinstance(distribution, distributions.DiscreteUniformDistribution):
+                value = value * distribution.q + distribution.low
+
+            params[name] = distribution.to_internal_repr(value)
+
+        return params
+
+    def _is_compatible(self, trial):
```
Comment (toshihikoyanase) [inline optuna/integration/skopt.py]: Currently, the test code does not cover the case `_is_compatible(trial) == False`.
Could you add the test cases to confirm that this method is able to filter out incompatible trials?

File: tests/integration_tests/test_skopt.py (hunk 6 above)
Comment (toshihikoyanase) [reply tests/integration_tests/test_skopt.py]: You're right. Thanks.

File: optuna/integration/skopt.py (hunk 7 above)
Comment (sile) [reply optuna/integration/skopt.py]: Added: https://github.com/pfnet/optuna/pull/399/commits/9a2519bd6683a9f3bd902ec22617097431603800

File: optuna/integration/skopt.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    Example:
+
+        Optimize a simple quadratic function by using :class:`~optuna.integration.SkoptSampler`.
+
+        .. code::
+
+                def objective(trial):
+                    x = trial.suggest_uniform('x', -100, 100)
+                    y = trial.suggest_int('y', -10, 10)
+                    return x**2 + y
```
Comment (toshihikoyanase) [inline optuna/integration/skopt.py]: I think the search space seems too wide compared with the number of trials.
When I executed the current code three times, the average of the best values is about 10. The error seems big, and users may think that the search efficiency of SkoptSampler is not so good.

What do you think of using a narrower search space?
For instance, the following setting marked `2.1e-9` as the average of the best score.

```python
def objective(trial):
    x = trial.suggest_uniform('x', -10, 10)
    y = trial.suggest_int('y', 0, 10)
    return x**2 + y
```

Comment (sile) [reply optuna/integration/skopt.py]: Sounds good. Thank you for your suggestion.

Fixed: https://github.com/pfnet/optuna/pull/399/commits/fe3279e9447254e71ea5f8edf25e18af9044aca1

File: optuna/integration/skopt.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+    def ask(self):
+        # type: () -> Dict[str, float]
+
+        params = {}
+        param_values = self._optimizer.ask()
+        for (name, distribution), value in zip(sorted(self._search_space.items()), param_values):
+            if isinstance(distribution, distributions.DiscreteUniformDistribution):
+                value = value * distribution.q + distribution.low
+
+            params[name] = distribution.to_internal_repr(value)
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/integration/skopt.py]: 📝

File: optuna/testing/sampler.py [hunk 10]
```diff
@@ -44,3 +44,22 @@ def sample_independent(self, study, trial, param_name, param_distribution):
...
 
         return param_distribution.to_internal_repr(param_value)
+
+
+class FirstTrialOnlyRandomSampler(optuna.samplers.RandomSampler):
+    def sample_relative(self, study, trial, search_space):
+        # type: (InTrialStudy, FrozenTrial, Dict[str, BaseDistribution]) -> Dict[str, float]
+
+        if len(study.trials) > 1:
+            raise RuntimeError("`FirstTrialOnlyRandomSampler` only works on the first trial")
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/testing/sampler.py]: Thanks. Added: https://github.com/pfnet/optuna/pull/399/commits/935db9745bae75568f92a315768048ff5037591c

File: optuna/integration/skopt.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+class _Optimizer(object):
+    def __init__(self, search_space, skopt_kwargs):
+        # type: (Dict[str, BaseDistribution], Dict[str, Any]) -> None
+
+        self._search_space = search_space
+
+        dimensions = []
+        for name, distribution in sorted(self._search_space.items()):
+            if isinstance(distribution, distributions.UniformDistribution):
+                high = max(distribution.low, np.nextafter(distribution.high, float('-inf')))
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/integration/skopt.py]: Yes. Your understanding is correct.

I added comments (and removed unnecessary `max` calls) at the commit https://github.com/pfnet/optuna/pull/399/commits/2ee3600f36ed66ac8efe2ecb616413a8f4c0709b.

File: optuna/integration/skopt.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        # type: (InTrialStudy, FrozenTrial) -> Dict[str, BaseDistribution]
+
+        search_space = {}
+        for name, distribution in samplers.product_search_space(study).items():
+            if distribution.single():
+                if not isinstance(distribution, distributions.CategoricalDistribution):
+                    # `skopt` cannot handle non-categorical distributions that contain just
+                    # a single value, so we skip this distribution.
+                    #
+                    # Note that `Trial` takes care of this distribution when suggestion.
```
Comment (g-votte) [inline optuna/integration/skopt.py]: Please rewrite this sentence like: Note that `Trial` takes care of this distribution during suggestion.

Comment (sile) [reply optuna/integration/skopt.py]: Thanks. Fixed: https://github.com/pfnet/optuna/pull/399/commits/86551d667b0473d5a669d7d4b3c6981d5f5b38ae

File: optuna/integration/skopt.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sile) [reply optuna/integration/skopt.py]: Fixed: https://github.com/pfnet/optuna/pull/441/commits/49b25a565bdd96aa67af09d458c42eb691858fd2
